src/test/java/org/datadog/jenkins/plugins/datadog/clients/DatadogClientStub.java

```diff
@@ -372,24 +257,103 @@ public static Map<String, Set<String>> addTagToMap(Map<String, Set<String>> tags
         return tags;
     }
 
+    private final StubTraceWriteStrategy traceWriteStrategy = new StubTraceWriteStrategy();
+
+    private static final class StubTraceWriteStrategy implements TraceWriteStrategy {
+        private volatile boolean isWebhook = false;
+        private final Collection<TraceSpan> traces = new LinkedBlockingQueue<>();
+        private final Collection<JSONObject> webhooks = new LinkedBlockingQueue<>();
+
+        @Override
+        public JSONObject serialize(BuildData buildData, Run<?, ?> run) {
+            if (isWebhook) {
+                JSONObject json = new DatadogWebhookBuildLogic().finishBuildTrace(buildData, run);
+                webhooks.add(json);
+                return json;
+            } else {
+                TraceSpan span = new DatadogTraceBuildLogic().createSpan(buildData, run);
+                traces.add(span);
+                return new JsonTraceSpanMapper().map(span);
+            }
+        }
+
+        @Override
+        public Collection<JSONObject> serialize(FlowNode flowNode, Run<?, ?> run) {
+            if (isWebhook) {
+                Collection<JSONObject> spans = new DatadogWebhookPipelineLogic().execute(flowNode, run);
+                webhooks.addAll(spans);
+                return spans;
+            } else {
+                Collection<TraceSpan> traceSpans = new DatadogTracePipelineLogic().collectTraces(flowNode, run);
+                traces.addAll(traceSpans);
+                JsonTraceSpanMapper mapper = new JsonTraceSpanMapper();
+                return traceSpans.stream().map(mapper::map).collect(Collectors.toList());
+            }
+        }
+
+        @Override
+        public void send(List<JSONObject> spans) {
+            // no op
+        }
+
+        public void configureForWebhooks() {
+            isWebhook = true;
+        }
+    }
+
     public void configureForWebhooks() {
-        traceBuildLogic = new DatadogWebhookBuildLogic(this);
-        tracePipelineLogic = new DatadogWebhookPipelineLogic(this);
+        traceWriteStrategy.configureForWebhooks();
+    }
+
+    @Override
+    public TraceWriteStrategy createTraceWriteStrategy() {
+        return traceWriteStrategy;
     }
 
     public boolean waitForWebhooks(final int number) throws InterruptedException {
-        final CountDownLatch latch = new CountDownLatch(number);
-        synchronized (webhookLatches) {
-            if (webhooks.size() >= number) {
+        long timeout = System.currentTimeMillis() + TimeUnit.SECONDS.toMillis(10);
+        while (System.currentTimeMillis() < timeout) {
+            if (traceWriteStrategy.webhooks.size() >= number) {
                 return true;
             }
-            webhookLatches.add(latch);
+            Thread.sleep(100L);
+        }
```

Comment:
Any particular reason why we have changed this? (Use fixed Thread.sleep instead of CountDownLatch)

Reply:
It just felt unnecessarily complicated and error-prone: if all the webhooks were there the latch was not created, otherwise it was created with the required total number - not considering the webhooks that were already in the queue.
I'd have to fix that, and replicated the same behaviour for waiting for the traces.
Given that these are tests, I decided that busy-waiting was acceptable - the logic is simpler, we don't burn that much CPU because of the sleep calls, the max time that we can waste waiting is 100ms because of the polling period.